repo/repo.go

```diff
@@ -71,72 +72,102 @@ func FindExternalRepo(repoRoot, name string) (string, error) {
 func ListRepositories(workspace *rule.File) (repos []*rule.Rule, repoFileMap map[string]*rule.File, err error) {
 	repoIndexMap := make(map[string]int)
 	repoFileMap = make(map[string]*rule.File)
-	for _, repo := range workspace.Rules {
-		if name := repo.Name(); name != "" {
-			repos = append(repos, repo)
-			repoFileMap[name] = workspace
-			repoIndexMap[name] = len(repos) - 1
-		}
-	}
-	extraRepos, err := parseRepositoryDirectives(workspace.Directives)
+
+	defs, err := parseRepositoryDirectives(workspace, true)
 	if err != nil {
 		return nil, nil, err
 	}
-	for _, repo := range extraRepos {
-		if i, ok := repoIndexMap[repo.Name()]; ok {
-			repos[i] = repo
-		} else {
+
+	for _, r := range defs {
+		repo := r.repo
+		file := r.file
+
+		if repo.Name() == "" {
+			continue
+		}
+		name := repo.Name()
+		i, ok := repoIndexMap[name]
+		if !ok {
+			// This name has not been seen before, append it
 			repos = append(repos, repo)
+			repoFileMap[name] = file
+			repoIndexMap[name] = len(repos) - 1
+			continue
 		}
-		repoFileMap[repo.Name()] = workspace
+
+		// We have a duplicate declaration of this name. The rest of this logic determines which one we want to keep.
+		other := repos[i]
+
+		if isFromDirective(other) && !isFromDirective(repo) {
+			// Directives are always preferred
+			continue
+		}
+
+		if repoFileMap[name] == file && repo.Chunk() > 0 && other.Chunk() > 0 {
+			// These are two rule definitions from the same chunked file
+			if other.Chunk() < repo.Chunk() {
+				// The existing definition occurred in an earlier chunk, and has therefore been frozen.
+				// Bazel will ignore the new definition, so we will too.
+				continue
+			}
+		}
+
+		// Because parseRepositoryDirectives preserves ordering, repo comes after other, so it overrides it.
+		repos[i] = repo
+		repoFileMap[name] = file
+	}
+
+	return repos, repoFileMap, nil
+}
+
+type repoDef struct {
+	repo *rule.Rule
+	file *rule.File
+}
+
+func isFromDirective(repo *rule.Rule) bool {
+	b, ok := repo.PrivateAttr(config.GazelleFromDirectiveKey).(bool)
+	if !ok {
+		return false
+	}
+	return b
+}
+
+func parseRepositoryDirectives(file *rule.File, recursive bool) (repos []repoDef, err error) {
```

Comment:
I'd rather not support recursion here. It's just more complexity than I'm willing to deal with.

Reply:
This code has changed beyond recognition